spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CategoricalColVocaList.scala

```diff
@@ -0,0 +1,134 @@
+ *
+ * By default, out-of-vocabulary values are ignored.
+ * Use either (but not both) of num_oov_buckets and default_value
+ * to specify how to include out-of-vocabulary values.
+ *
+ * if isSetDefault enabled, the defalut value len(vocabulary_list) will be set.
+ *
+ * if num_oov_buckets enabled, all out-of-vocabulary inputs will be assigned IDs in the range
+ * [len(vocabulary_list), len(vocabulary_list)+num_oov_buckets) based on a hash of the input value
+ *
+ * A positive num_oov_buckets can not be specified with default_value.
+ *
+ * the input Tensor[String] can be 1-D or 2-D Tensor.
+ *
+ * The Operation support the feature column with single-value or multi-value
+ *
+ * The missing values in input Tensor can be represented by -1 for int and '''' for string
+ *
+ * @param vocaList An vocabulary with the length more than or equal to 1.
+ * @param strDelimiter The delimiter of feature string, default: ",".
+ * @param isSetDefault Set default value for out-of-vocabulary feature values default: false.
+ * @param numOovBuckets the non-negative number of out-of-vocabulary buckets, default: 0.
+ * @tparam T Numeric type. Parameter tensor numeric type. Only support float/double now
+ */
+
+class CategoricalColVocaList[T: ClassTag](
+  val vocaList: Array[String],
+  val strDelimiter: String = ",",
+  val isSetDefault: Boolean = false,
+  val numOovBuckets: Int = 0
+) (implicit ev: TensorNumeric[T])
+  extends Operation[Tensor[String], Tensor[Int], T]{
+
+  private val vocaLen = vocaList.length
+  private val vocaMap = vocaList.zipWithIndex.toMap
+
+  require(numOovBuckets >= 0,
+    "numOovBuckets is a negative integer")
+  require(!(isSetDefault && numOovBuckets != 0),
+    "defaultValue and numOovBuckets are both specified")
+  require(vocaLen > 0,
+    "the vocabulary list is empty")
+  require(vocaLen == vocaMap.size,
+    "the vocabulary list contains duplicate keys")
+
+  output = Tensor[Int]()
+
+  override def updateOutput(input: Tensor[String]): Tensor[Int] = {
+
+    input.squeeze()
+    val rows = input.size(dim = 1)
+
+    val cols = if (numOovBuckets==0) {
+      if (isSetDefault) vocaLen + 1 else vocaLen
+    }
+    else {
+      vocaLen + numOovBuckets
+    }
+    val shape = Array(rows, cols)
+    val indices0 = new ArrayBuffer[Int]()
+    val indices1 = new ArrayBuffer[Int]()
+    val values = new ArrayBuffer[Int]()
+
+    var i = 1
+    while (i <= rows) {
+      var feaStrArr = input.valueAt(i).split(strDelimiter)
+      if (!isSetDefault && numOovBuckets == 0) {
+        feaStrArr = feaStrArr.filter(x => vocaMap.contains(x))
+      }
+      var j = 0
+      while (j < feaStrArr.length) {
+        val mapVal = numOovBuckets==0 match {
+          case true =>
+            vocaMap.getOrElse(feaStrArr(j), vocaMap.size)
+          case false =>
+            vocaMap.getOrElse(feaStrArr(j),
+              MurmurHash3.stringHash(feaStrArr(j)) % numOovBuckets match {
```

Comment:
Have seen this in another PR. Maybe consider create a util method for this function?

Reply:
a good idea! but i'm not sure where to create the util method, `nn.ops.*` or `nn.utils.Util.scala`?

Reply:
```nn.Utils``` should be fine